### PR TITLE
STITCH-873 Create Client API v2.0 for iOS SDK

### PR DESCRIPTION
**Pull request by jsflax**

### CODE REVIEW

File: StitchLogger/StitchLogger/Source/LogManager.swift [hunk 1]
```diff
@@ -19,9 +19,9 @@ public enum LogLevel: Int, CustomStringConvertible {
         case .info:
```
Comment (edaniels) [inline StitchLogger/StitchLogger/Source/LogManager.swift]: [nit] Let's get rid of all emojis in this file. Including the leaves

File: StitchCore/StitchCoreTests/AuthTests.swift [hunk 2]
```diff
@@ -0,0 +1,85 @@
+//
+//  AuthTests.swift
+//  StitchCoreTests
+//
+//  Created by Jason Flax on 11/18/17.
```
Comment (edaniels) [inline StitchCore/StitchCoreTests/AuthTests.swift]: [nit] Remove these auto generated "Created by"s in each file.

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 3]
```diff
@@ -6,7 +6,7 @@ import Security
 
 public struct Consts {
     public static let DefaultBaseUrl =   "https://stitch.mongodb.com"
-    static let ApiPath =                 "/api/client/v1.0/app/"
+    static let ApiPathV2 =               "/api/client/v2.0/"
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [optional] Doesn't seem necessary to suffix V2 (for now)

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 4]
```diff
@@ -23,84 +23,95 @@ public struct Consts {
...
-    static let PushPath =                "push"
 }
 
 /// A StitchClient is responsible for handling the overall interaction with all Stitch services.
 public class StitchClient: StitchClientType {
     // MARK: - Properties
     /// Id of the current application
     public var appId: String
-
+    internal lazy var routes = Routes(appId: appId)
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [😎]

File: StitchCore/StitchCore/Source/Auth/AuthProviderInfo.swift [hunk 5]
```diff
@@ -1,14 +1,101 @@
...
-    private enum CodingKeys: String, CodingKey {
-        case anonymousAuthProviderInfo = "anon/user"
-        case googleProviderInfo = "oauth2/google"
-        case facebookProviderInfo = "oauth2/facebook"
-        case emailPasswordAuthProviderInfo = "local/userpass"
+public protocol AuthProviderType: Codable {
+    var id: String { get }
+    var name: String { get }
+    var type: String { get }
+    var disabled: Bool { get }
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Auth/AuthProviderInfo.swift]: [whoops] I'm going to make this not show to clients. You should only ever see enabled providers anyway. You will only get name, type, metadata_fields, and config

File: StitchCore/StitchCore/Source/Auth/Providers/Anonymous/AnonymousAuthProvider.swift [hunk 6]
```diff
@@ -3,10 +3,10 @@ import ExtendedJson
 /// AnonymousAuthProvider provides a way to authenticate anonymously.
 public struct AnonymousAuthProvider: AuthProvider {
     /// The authentication type for anonymous login.
-    public let type: String = "anon"
+    public let type: String = "anon-user"
 
     /// The name for anonymous login
-    public let name: String = "user"
+    public let name: String = "anon-user"
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Auth/Providers/Anonymous/AnonymousAuthProvider.swift]: [minor] The name is technically configurable. The name if the response from auth providers is the name you use in the URL, not the type.

File: StitchCore/StitchCore/Source/Auth/Providers/EmailPassword/EmailPasswordAuthProvider.swift [hunk 7]
```diff
@@ -5,12 +5,12 @@ import ExtendedJson
...
     /// The authentication type for email/pass login.
     public var type: String {
-        return "local"
+        return "local-userpass"
     }
 
     /// The name for email/pass login
     public var name: String {
-        return "userpass"
+        return "local-userpass"
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Auth/Providers/EmailPassword/EmailPasswordAuthProvider.swift]: Same name comment for all providers

File: StitchCore/StitchCore/Source/Core/Functions.swift [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+//  Functions.swift
+//  StitchCore
+//
+//  Created by Jason Flax on 11/14/17.
+//  Copyright © 2017 MongoDB. All rights reserved.
+//
+
+import ExtendedJson
+
+internal struct Function: Encodable {
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/Functions.swift]: [nit] I don't think this name conveys what this `struct` is. In the server we call this a `FunctionCallRequest`

File: StitchCore/StitchCore/Source/Core/Services/Services.swift [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+protocol Service {
+    var client: StitchClient { get }
+    var name: String { get }
+}
+
+public class Services {
+    let client: StitchClient
+
+    func twilio(name: String) -> TwilioService {
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/Services/Services.swift]: [q] Are you okay with this kind of coupling?

File: StitchCore/StitchCore/Source/Core/StitchClient.swift [hunk 10]
```diff
@@ -161,11 +177,12 @@ public class StitchClient: StitchClientType {
...
     @discardableResult
     public func emailConfirm(token: String, tokenId: String) -> StitchTask<Void> {
-        return self.performRequest(method: .post,
-                                   endpoint: Consts.AuthPath + "/local/userpass/confirm",
-                                   isAuthenticatedRequest: false,
-                                   parameters: ["token": token, "tokenId": tokenId],
-                                   responseType: [String: String].self).then { _ in return }
+        return httpClient.doRequest {
+            $0.method = .post
+            $0.endpoint = self.routes.authProvidersExtensionRoute + "/local-userpass/confirm"
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/StitchClient.swift]: [style] local-userpass should be extracted into a constant

File: StitchCore/StitchCore/Source/Core/StitchHTTPClient.swift [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+                        task.result = .success(json)
+                    case .failure(let error):
+                        task.result = .failure(error)
+                    }
+            })
+
+        return task
+    }
+
+    internal func parseErrorV2(from value: [String: Any]) -> StitchError? {
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/StitchHTTPClient.swift]: [nit] Do not need v2

File: StitchCore/StitchCore/Source/Core/StitchHTTPClient.swift [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+                            task.result = .failure(error)
+                        case .success(let value):
+                            task.result = .success(value)
+                        }
+                }
+            }
+        }
+    }
+
+    private func urlV2(withEndpoint endpoint: String) -> String {
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Core/StitchHTTPClient.swift]: [nit] Don't need v2

File: StitchCore/StitchCore/Source/Network/Adapter/StitchNetworkAdapter.swift [hunk 13]
```diff
@@ -18,22 +18,29 @@ public class StitchNetworkAdapter: NetworkAdapter {
...
         let task = StitchTask<Data?>()
-        let defaultSession = URLSession(configuration: .default)
+
+        let config = URLSessionConfiguration.default
+        config.requestCachePolicy = .reloadIgnoringLocalCacheData
+
+        let defaultSession = URLSession(configuration: config)
 
         guard let url = URL(string: url) else {
+            printLog(.error, text: "bad url")
```
Comment (edaniels) [inline StitchCore/StitchCore/Source/Network/Adapter/StitchNetworkAdapter.swift]: [q] Log necessary here?

File: StitchCore/StitchCoreTests/ServiceTests.swift [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+//  Copyright © 2017 MongoDB. All rights reserved.
+//
+import XCTest
+import Foundation
+import StitchLogger
+import ExtendedJson
+@testable import StitchCore
+
+class ServiceTests: XCTestCase {
+    let client = StitchClient(appId: "test-jsf-fpleb", baseUrl: "https://stitch-dev.mongodb.com")
```
Comment (edaniels) [inline StitchCore/StitchCoreTests/ServiceTests.swift]: [nit] Should either make this configurable or use prod

File: StitchCore/StitchCoreTests/StitchCoreTests.swift [hunk 15]
```diff
@@ -15,7 +15,7 @@ class StitchCoreTests: XCTestCase {
         super.tearDown()
     }
 
-    let stitchClient = StitchClient(appId: "test-uybga")
+    let stitchClient = StitchClient(appId: "test-jsf-fpleb", baseUrl: "https://stitch-dev.mongodb.com")
```
Comment (edaniels) [inline StitchCore/StitchCoreTests/StitchCoreTests.swift]: ibid

File: StitchCore/StitchCore/Source/Core/Services/Services.swift (hunk 9 above)
Comment (jsflax) [reply StitchCore/StitchCore/Source/Core/Services/Services.swift]: I'm going to hide this behind a private flag for now.

**Comment by jsflax** [comment]
@edaniels PTAL

**Pull request merged**